backends/ebpf/psa/externs/ebpfPsaMeter.cpp

```diff
@@ -0,0 +1,435 @@
+    }
+}
+
+IR::IndexedVector<IR::StructField> EBPFMeterPSA::getValueFields() {
+    auto vec = IR::IndexedVector<IR::StructField>();
+    auto bits_64 = new IR::Type_Bits(64, false);
+    vec.push_back(new IR::StructField(IR::ID("pir_period"), bits_64));
+    vec.push_back(new IR::StructField(IR::ID("pir_unit_per_period"), bits_64));
+    vec.push_back(new IR::StructField(IR::ID("cir_period"), bits_64));
+    vec.push_back(new IR::StructField(IR::ID("cir_unit_per_period"), bits_64));
+    vec.push_back(new IR::StructField(IR::ID("pbs"), bits_64));
+    vec.push_back(new IR::StructField(IR::ID("cbs"), bits_64));
+    vec.push_back(new IR::StructField(IR::ID("pbs_left"), bits_64));
+    vec.push_back(new IR::StructField(IR::ID("cbs_left"), bits_64));
+    vec.push_back(new IR::StructField(IR::ID("time_p"), bits_64));
+    vec.push_back(new IR::StructField(IR::ID("time_c"), bits_64));
+    return vec;
+}
+
+IR::Type_Struct *EBPFMeterPSA::createSpinlockStruct() {
+    auto spinLock = new IR::Type_Struct(IR::ID("bpf_spin_lock"));
+    return spinLock;
+}
+
+void EBPFMeterPSA::emitSpinLockField(CodeBuilder* builder) const {
+    auto spinlockStruct = createSpinlockStruct();
+    auto spinlockType = EBPFTypeFactory::instance->create(spinlockStruct);
+    builder->emitIndent();
+    spinlockType->declare(builder, spinlockField, false);
+    builder->endOfStatement(true);
+}
+
+void EBPFMeterPSA::emitKeyType(CodeBuilder* builder) const {
+    builder->emitIndent();
+    builder->append("typedef ");
+    this->keyType->declare(builder, keyTypeName, false);
+    builder->endOfStatement(true);
+}
+
+void EBPFMeterPSA::emitValueStruct(CodeBuilder* builder, P4::ReferenceMap* refMap) {
+    builder->emitIndent();
+    getBaseValueType(refMap)->emit(builder);
+}
+void EBPFMeterPSA::emitValueType(CodeBuilder* builder) const {
+    if (isDirect) {
+        builder->emitIndent();
+        getBaseValueType(program->refMap)->declare(builder, instanceName, false);
+        builder->endOfStatement(true);
+    } else {
+        getIndirectValueType()->emit(builder);
+    }
+}
+
+void EBPFMeterPSA::emitInstance(CodeBuilder *builder) const {
+    if (!isDirect) {
+        builder->target->emitTableDeclSpinlock(builder, instanceName, TableHash,
+                                               this->keyTypeName,
+                                               "struct " + getIndirectStructName(), size);
+    } else {
+        ::error(ErrorType::ERR_UNEXPECTED, "Direct meter belongs to table "
+                                           "and cannot have own instance");
+    }
+}
+
+void EBPFMeterPSA::emitExecute(CodeBuilder* builder, const P4::ExternMethod* method,
+                               ControlBodyTranslatorPSA* translator) const {
+    auto pipeline = dynamic_cast<const EBPFPipeline *>(program);
+    if (pipeline == nullptr) {
+        ::error(ErrorType::ERR_INVALID, "Meter used outside of pipeline %1%", method->expr);
```

Comment:
Is "pipeline" an ebpf concept? It is not a P4 concept.

Reply:
@mbudiu-vmw We have introduced a concept of "pipeline" in the PSA/eBPF backend to distinguish between EBPF program used by ebpf_model and a new PSA arch which has two pipelines. PSA defines ingress and egress pipelines so we should have two EBPFPipeline objects representing them.